### PR TITLE
BinaryReader over Uint8Array (#11)

### LINKED ISSUE #11: BinaryReader over Uint8Array
**Package:** `packages/core` (npm `unity-asset-reader`)

## Description
`BinaryReader` directly over `Uint8Array`. **No `ByteSource` interface yet** (nor anywhere in the repo) - one impl and its only other consumer is a backlog item; the seam is introduced in #51 when it triggers.

## Tasks
- [ ] `BinaryReader`: endian switch, u8/i8/u16/i16/u32/i32/f32/f64, **u64/i64 as `bigint`** (plan D9), aligned read, cstring, length-prefixed string, `align(n)`, bytes (zero-copy `subarray`, never `slice`)
- [ ] Guard: offsets > 2^53 throw
- [ ] Unit tests incl. both endians, alignment, bigint edges

## Port
`AssetStudio/EndianBinaryReader.cs` (361 LOC)

---
Plan: [`docs/unity-asset-reader-plan.md`](../blob/main/docs/unity-asset-reader-plan.md). Behavior source of truth: [Razviar/assetstudio](https://github.com/Razviar/assetstudio) (MIT). Golden oracle: [UnityPy](https://github.com/K0lb3/UnityPy) (MIT). **MIT-derivative port: keep upstream copyright notices. Do not copy or import `@arkntools/unity-js` (AGPL-3.0), tests included. No C# compiled to WASM - C# is reference only, hand-ported to TS; WASM is built from C/C++ only (plan D2).** Layout: npm-workspaces monorepo under `packages/`; dependencies point feature package -> core only (plan D7/D8, rules R13/R14).

### PR DESCRIPTION
**Pull request by fatal10110**
Closes #11

## What

`BinaryReader` in `packages/core/src/io/BinaryReader.ts`, a cursor over a `Uint8Array` ported from `AssetStudio/EndianBinaryReader.cs`. Switchable endianness (`endian` is a writable property because `SerializedFile` flips it mid-file from `m_Endianess`), the 8/16/32-bit integer and float reads, 64-bit integers as `bigint` (D9), `align(n)`, Unity's length-prefixed string and NUL-terminated C string, and `readBytes` returning zero-copy `subarray` views (R7).

Two decisions worth knowing:

- **Offsets stay `number`, with a hard stop at 2^53** (D9). `position` and `readBytes` split their failures the same way: a `RangeError` when the value is not a usable offset or size at all (negative, fractional, 2^53 and above — nothing an integer read can produce, so it is a caller mistake), and a `CorruptError` only when the number is fine but the bytes are not there.
- **No `ByteSource` interface** — one implementation, its only other consumer is a backlog item, and the seam is #51's job.

UTF-8 decoding uses the WHATWG `TextDecoder`, declared locally in the file: the package compiles with `lib: ["ES2020"]` and `types: []`, so nothing declares it, and widening `lib` to DOM would defeat the R4 guard. `TextDecoder` is a platform global in browsers, workers, Deno and Node ≥ 11, so this stays isomorphic. The local `declare` is module-scoped and does not leak into `dist/io/BinaryReader.d.ts`.

## Acceptance

The issue lists Tasks rather than Acceptance bullets; each Task bullet is mapped below.

| Criterion (from the issue) | Proof |
|---|---|
| endian switch | `tests/BinaryReader.test.ts` › "defaults to big-endian, like upstream EndianBinaryReader" · "the endian switch applies to later reads (SerializedFile flips mid-file)" · "single bytes ignore the endian setting" |
| u8/i8/u16/i16/u32/i32/f32/f64 | `tests/BinaryReader.test.ts` › "reads every fixed-width type big-endian" · "reads every fixed-width type little-endian" |
| **u64/i64 as `bigint`** (plan D9) | `tests/BinaryReader.test.ts` › "64-bit integers come back as bigint in both endians (D9)" (each endian reads i64 max, i64 min and −2, so the signed path is load-bearing in both) · "64-bit edge values survive exactly" (u64 max, i64 min, −1, 0) · "bigint keeps precision a number would lose above 2^53" |
| aligned read | `tests/BinaryReader.test.ts` › "reads a length-prefixed string and the padding after it" · "a length-prefixed string needing no padding does not consume any" · "aligning after a length-prefixed string never runs past the end" |
| cstring | `tests/BinaryReader.test.ts` › "reads a NUL-terminated C string and consumes the terminator" · "a C string decodes multi-byte UTF-8" · "an empty C string consumes only its terminator" · "a C string stops at maxLength without consuming a terminator" · "an unterminated C string stops at the end, like upstream" |
| length-prefixed string | `tests/BinaryReader.test.ts` › "reads a length-prefixed string and the padding after it" · "a length-prefixed string honours the endian switch" · "an empty or impossible length yields \"\" and still aligns, as upstream does" |
| `align(n)` | `tests/BinaryReader.test.ts` › "align(n) skips to the next boundary and is a no-op when already aligned" · "align stops at the end instead of running past it" · "align rejects a non-positive alignment" |
| bytes (zero-copy `subarray`, never `slice`) | `tests/BinaryReader.test.ts` › "readBytes returns a zero-copy view aliasing the input (R7)" (asserts same `buffer`, same `byteOffset`, and that writes alias both ways) · "readBytes(0) returns an empty view and does not move" · "reads through a view into a larger buffer without seeing its neighbours" |
| Guard: offsets > 2^53 throw | `tests/BinaryReader.test.ts` › "an offset above 2^53 throws rather than losing precision (D9)" · "an offset that is not a usable number is the caller's mistake, not corrupt data" · "an offset past the end is corrupt data, not a RangeError" · "readBytes rejects a count that is not a whole number in 0..2^53" |
| Unit tests incl. both endians, alignment, bigint edges | 30 tests in `packages/core/tests/BinaryReader.test.ts`, plus the unhappy path: "reading past the end throws CorruptError naming what was short" |

## Verification

Run in a git worktree, so **root `npm run verify` was not run**: it builds `texture2ddecoder-wasm`, whose `wasm/` artifact is gitignored, absent from a fresh checkout, and needs Docker to rebuild. `packages/texture2ddecoder-wasm/` is untouched by this PR (R13). The pieces of `verify` that cover this change were run instead:

```
npm run build -w unity-asset-reader   ✅
npm test  -w unity-asset-reader       ✅  45 passed, 0 failed
npm run check:browser                 ✅  core, texture, texture2ddecoder-wasm
npm run check:no-csharp               ✅
```

CI runs the full root `verify`.

## Reference

Ported from: `AssetStudio/EndianBinaryReader.cs` (`ReadInt16`…`ReadDouble`, `ReadBytes`, `AlignStream`, `ReadAlignedString`, `ReadStringToNull`). Read at `Razviar/assetstudio@main`, cloned outside the repo (R2). `@arkntools/unity-js` was not consulted (R1).

Differences from upstream, all deliberate:

| Upstream | Here | Why |
|---|---|---|
| `ReadBytes` copies into a fresh `byte[]` and returns **short** at EOF (`if (n == 0) break;`) | `readBytes` returns a `subarray` view and throws `CorruptError` when fewer than `count` bytes remain | R7 (no copies on the read path) and R9 (never hand back partial data silently). A short read upstream surfaces later as a confusing parse failure. |
| `AlignStream` can leave `Stream.Position` past `Length` | `align(n)` stops at the end of the data | A `Uint8Array` cursor has nowhere past the end to point, and there is no padding there. The next read still reports the truncation. |
| `Position` accepts any `long` | `position` throws: `RangeError` below 0, fractional, or 2^53 and above; `CorruptError` past `length` | D9. |
| `ReadQuaternion`/`ReadVector2`/`ReadMatrix`/`ReadMhy*`/the `Read*Array` family | not ported | Out of scope for this issue (R7 scope rule); the math types they need do not exist yet, and the `Read*Array` helpers belong with the typetree reader that calls them. |
| `EndianType` enum | `type Endian = "big" \| "little"` | No enum needed in TS; a string union reads better at call sites and survives `isolatedModules`. |

`ReadAlignedString`'s tolerance is kept as-is: a length that is ≤ 0 or longer than what remains yields `""` and still aligns, because Unity writes such lengths for empty fields and upstream's behavior on those bytes is what we have to match.

## Deviations

- **Errors.** The task brief assumed `packages/core/src/errors.ts` was still unmerged and suggested a plain `RangeError`. #64 has since merged, so data-driven failures — a read or a seek that lands past the end — throw `CorruptError` with the offset and the shortfall in the message, per the "errors carry context" rule. `RangeError` covers arguments that are not usable offsets or sizes at all: negative, fractional, or 2^53 and above, plus `align(0)`. `position` and `readBytes` apply that split identically.
- **"aligned read" and "length-prefixed string"** are one method, `readAlignedString()`. In upstream they are one method too: `ReadAlignedString` reads an `Int32` length prefix, that many UTF-8 bytes, then aligns to 4. Adding a second, prefix-only string method with no caller would be the speculative API the rules forbid; it can be added by the issue that first needs it.
- `readBoolean` is not implemented — upstream inherits it from `System.IO.BinaryReader` and the issue does not list it. It is one line for whichever issue first reads a bool field.

## Follow-ups

None.

### CODE REVIEW

**Review by fatal10110 — COMMENTED** [review]
## Review: approved

Reviewed at `666c121` on a detached checkout of `pull/65/head`. Upstream read at `Razviar/assetstudio@main`, cloned outside the repo (R2).

### Verification (rerun by the reviewer)

Root `npm run verify` was not run, for the reason the PR body gives: it builds `texture2ddecoder-wasm`, whose `wasm/` artifact is gitignored and needs Docker. The parts of `verify` that cover this change were run instead, after `npm ci`:

| Command | Result |
|---|---|
| `npm run build -w unity-asset-reader` | pass |
| `npm test -w unity-asset-reader` | pass - 44 tests, 0 fail, 0 skipped, 0 todo |
| `npm run check:browser` | pass - core, texture, texture2ddecoder-wasm |
| `npm run check:no-csharp` | pass |
| `git ls-files '*.cs' '*.csproj' '*.sln'` (run by hand, not via the script) | prints nothing |

### Issue #11 tasks

The issue lists Tasks and no Acceptance bullets. One row per Task bullet, split where a bullet names several things.

| Task | Status | Proof |
|---|---|---|
| endian switch | **met** | `BinaryReader.ts:68-75`; tests "defaults to big-endian...", "the endian switch applies to later reads...", "single bytes ignore the endian setting" |
| u8/i8/u16/i16/u32/i32/f32/f64 | **met** | `BinaryReader.ts:110-165`; "reads every fixed-width type big-endian" / "...little-endian". Both use values that differ between byte orders (`1234`/`3412`, float 1.5, double -2.25), so a mis-threaded `little` flag in any one method fails the run. |
| u64/i64 as `bigint` (D9) | **met** | `BinaryReader.ts:144-155`; "64-bit integers come back as bigint in both endians (D9)", "64-bit edge values survive exactly" (2^64-1, -2^63, -1, 0), "bigint keeps precision a number would lose above 2^53". |
| aligned read | **met** | `readAlignedString()` at `BinaryReader.ts:217-225`. Checked against upstream: `EndianBinaryReader.cs:151-162` is a single `ReadAlignedString` that reads the `Int32` prefix, the bytes, then `AlignStream()`. There is no separate prefix-only string method on upstream's reader (`WebFile.cs:39-41` does its unaligned length-prefixed read inline with `ReadInt32` + `ReadBytes`, not through a reader method). Merging the two bullets is therefore a faithful port, not a narrowing. |
| cstring | **met** | `readStringToNull()` at `BinaryReader.ts:236-244`; five tests incl. multi-byte UTF-8, `maxLength` cutoff and the unterminated case. Behavior matches `EndianBinaryReader.cs:164-179` on all four paths: terminator consumed when found; `maxLength` bytes of text, terminator not consumed; a NUL sitting exactly at `maxLength` not consumed; unterminated stops at the end. |
| length-prefixed string | **met** | same method, see "aligned read". |
| `align(n)` | **met** | `BinaryReader.ts:197-205`; three tests incl. the default-4 overload and the non-positive-alignment rejection. |
| bytes, zero-copy `subarray`, never `slice` (R7) | **met** | `BinaryReader.ts:183`; "readBytes returns a zero-copy view aliasing the input (R7)" asserts same `buffer` **and** same `byteOffset` **and** two-way write-through, so swapping `subarray` for `slice` fails it. "reads through a view into a larger buffer..." additionally pins `byteOffset` for a reader built on a mid-buffer view. |
| Guard: offsets > 2^53 throw | **met** | `BinaryReader.ts:87-97`; "an offset above 2^53 throws rather than losing precision (D9)" sweeps 2^53, 2^53+2, 2^60, `MAX_VALUE`, `Infinity` and asserts the cursor is untouched afterwards. |
| Unit tests incl. both endians, alignment, bigint edges | **met** | 29 tests in `packages/core/tests/BinaryReader.test.ts`, incl. the unhappy path "reading past the end throws CorruptError naming what was short". |

### Rules

| ID | Result |
|---|---|
| R1 | Clean. `grep -rni arkntools` over the tree hits only the rules/plan/AGENTS docs, never code or tests. |
| R2 | Guard prints nothing; no `.cs` added. |
| R3 | `BinaryReader.ts:1` carries the exact attribution line for `AssetStudio/EndianBinaryReader.cs`. `packages/core/NOTICE` already lists AssetStudio and UnityPy, so no NOTICE change was needed. |
| R4 | No `node:` / `fs` / `Buffer` / `process` / `__dirname` / DOM in `packages/core/src`. See the `TextDecoder` note below. |
| R5 | No `async` / `await` / `Promise` anywhere in `packages/core/src`. |
| R6 | `readUInt64` / `readInt64` return `bigint` and are never funnelled through `Number()`. Offsets stay `number` with the 2^53 throw. |
| R7 | No `.slice(` in `packages/core/src`. |
| R8 | No dependency change - `package.json` untouched in the diff. |
| R9 | Short reads and out-of-range offsets throw `CorruptError` with offset and shortfall; nothing returns partial data. |
| R13 | `packages/texture2ddecoder-wasm/` untouched. |
| R14 | Core imports no workspace package; `check:browser` enforces it and passes. |
| Scope | Exactly three files: `src/io/BinaryReader.ts` (replacing the `export {}` placeholder), `tests/BinaryReader.test.ts`, and one export line in `src/index.ts`. No `ByteSource` interface - correctly deferred to #51. |

### The declared divergences from upstream, checked

Each was checked against the upstream source rather than taken on the PR body's word. All four hold.

1. **`readBytes` throws instead of truncating.** Upstream `EndianBinaryReader.cs:110-134` breaks out on `n == 0` and returns short. No caller in this project's ported design depends on that: every `ReadBytes` in `BundleFile.cs` (lines 221, 277, 419, 424, 490, 551) and `SerializedFile.cs` (68, 231, 301, 305, 307, 398) takes a fixed size or a header-declared size where a short return means the file is broken - `BundleFile.cs:426` immediately does `blocksInfoBytes.AsSpan(0, compressedBlocksInfoSize)`, which throws in C# anyway. Throwing is the better port. See the inline note on `readBytes` for the one upstream site that *does* lean on short reads and is in M1 scope.
2. **`align(n)` clamps at the end.** Upstream `AlignStream` (`EndianBinaryReader.cs:141-149`) can push `Position` past `Length`. The four call sites in scope - `BundleFile.cs:413`, `BundleFile.cs:525`, `SerializedFile.cs:155`, `SerializedFile.cs:211` - all align mid-file with data still to come, and none reads `Position` afterwards as a value. Where upstream would overshoot, the next read throws either way, and clamping makes `remaining` stay non-negative. No behavioral loss.
3. **One `readAlignedString`.** Verified against upstream; see the task table.
4. **`readAlignedString` returns `""` on a non-positive or oversized length.** Verbatim: upstream is `if (length > 0 && length <= Remaining) { ... }` then `AlignStream()` unconditionally (`EndianBinaryReader.cs:153-161`). `Remaining` is evaluated after the prefix read there, as it is here. This is not an R9 silent-corruption case - it is the upstream contract for Unity's empty string fields, and the bytes the bad length claimed are deliberately left unread.
5. **Locally `declare`d `TextDecoder`.** Confirmed on three counts. (a) It does not leak: `packages/core/dist/io/BinaryReader.d.ts` contains no `TextDecoder`; the only hits in `dist/` are the runtime `new TextDecoder("utf-8")` in the two bundles. (b) It does not weaken the R4 guard: `scripts/check-browser.mjs` esbuild-bundles every source file for `platform: "browser"` and greps the output for `Buffer` / `process` / `__dirname`. A `declare` is erased at emit and cannot mask an unresolvable `node:` import or a node global, so the guard's reach is unchanged - the alternative (widening `lib` to DOM) is what would have weakened it. (c) `engines.node` is `>=18.0.0` and `TextDecoder` has been a global since Node 11, so it exists in every targeted runtime, and it is the WHATWG Encoding API rather than DOM. Default (non-fatal) construction gives U+FFFD replacement, matching `Encoding.UTF8.GetString`. `packages/core/src/env.ts` is not a globals module - plan §2 reserves it for the resource resolver - so there is no better home for the declaration.
6. **Omissions.** `ReadQuaternion` / `ReadVector2` / `ReadVector4` / `ReadColor4` / `ReadMatrix` / `ReadFloat` need math types that do not exist yet; `ReadMhyInt` / `ReadMhyUInt` / `ReadMhyString` are miHoYo-specific and outside the plan; the `Read*Array` family is a thin loop over the primitives that belongs with the typetree reader that calls it; `ReadBoolean` is inherited from `System.IO.BinaryReader`, not defined in the ported file. None appears in issue #11's Tasks. All correctly out of scope (R7 scope rule).

### Independent checks

- **bigint extremes.** 2^64-1, -2^63, -1 and 0 all round-trip exactly, and `9007199254740993n` is asserted to differ from `BigInt(Number(value))`, which pins the D9 claim rather than merely stating it.
- **Endian coverage.** Every multi-byte read is exercised in both byte orders with values that are not palindromic, including both floats. The one small gap is noted inline on the test file.
- **Zero-copy.** Covered, and the test genuinely fails on a `slice` swap.
- **`position`/`take` interaction.** A failed read leaves the cursor untouched (`take` throws before assigning), and that is asserted.

### Findings

No blockers and no majors. Three nits, all inline, none of which need to be addressed before merge.

File: packages/core/src/io/BinaryReader.ts [hunk 1]
```diff
@@ -1 +1,262 @@
...
+   * @returns a `subarray`, never a `slice`
+   * @throws {RangeError} when `count` is negative or not an integer below 2^53
+   * @throws {CorruptError} when fewer than `count` bytes remain
+   */
+  readBytes(count: number): Uint8Array {
+    if (!Number.isSafeInteger(count) || count < 0) {
+      throw new RangeError(`byte count ${count} is not an integer in 0..2^53`);
+    }
+    const start = this.take(count);
+    return this.bytes.subarray(start, start + count);
```
Comment (fatal10110) [inline packages/core/src/io/BinaryReader.ts:191]: **nit · note for a later issue, no change wanted here** Throwing instead of returning short is the right call for every `ReadBytes` site in `BundleFile.cs` and `SerializedFile.cs`, so this is correct as ported.

One upstream site does rely on the truncation, and it is in M1 scope (plan §107 lists `FileReader.cs` "detection only", i.e. the scaffolded `src/bundle/detect.ts`): `FileReader.CheckFileType` sniffs magics with `Position = 0x20; magic = ReadBytes(6);` and then `SequenceEqual`. On a file shorter than 0x26 bytes upstream gets a short array, the comparison is simply false, and detection falls through to the next candidate. Translated literally onto this reader, the `Position = 0x20` seek alone throws `CorruptError` on any file under 32 bytes, turning "not a Brotli file" into a parse failure.

Nothing to change in this PR - just make sure the detect port guards on `remaining` before each sniff rather than transcribing upstream's seek-and-compare.

File: packages/core/src/io/BinaryReader.ts [hunk 2]
```diff
@@ -1 +1,262 @@
...
+  }
+
+  set position(value: number) {
+    if (!Number.isSafeInteger(value)) {
+      throw new CorruptError(
+        `offset ${value} is not an integer below 2^53; offsets and sizes are numbers (D9)`,
+      );
+    }
+    if (value < 0 || value > this.bytes.length) {
+      throw new CorruptError(`offset ${value} is outside 0..${this.bytes.length}`);
```
Comment (fatal10110) [inline packages/core/src/io/BinaryReader.ts]: **nit** The PR body draws the line as "`RangeError` only for caller mistakes that have nothing to do with the data - `readBytes(-1)`, `align(0)`", but `position = -1` and `position = 1.5` are caller mistakes by that same test and get `CorruptError` here.

The out-of-range case is defensible (an offset usually comes from a header field that was parsed, so it *is* data-driven), but `1.5` cannot come from any integer read. Either fold the non-integer branch on line 88 into a `RangeError` and keep `CorruptError` for the range check, or drop the distinction and note in the JSDoc that offsets are treated as data-driven regardless of origin. Cosmetic - the thrown-vs-not behavior is right either way.

File: packages/core/tests/BinaryReader.test.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+});
+
+test("64-bit integers come back as bigint in both endians (D9)", () => {
+  const big = new BinaryReader(hex("00200000000000017fffffffffffffff"), "big");
+  assert.equal(big.readUInt64(), 9007199254740993n);
+  assert.equal(big.readInt64(), 9223372036854775807n);
+
+  const little = new BinaryReader(hex("0100000000002000ffffffffffffff7f"), "little");
+  assert.equal(little.readUInt64(), 9007199254740993n);
+  assert.equal(little.readInt64(), 9223372036854775807n);
```
Comment (fatal10110) [inline packages/core/tests/BinaryReader.test.ts:93]: **nit** This is the only little-endian `readInt64` assertion, and `7fffffffffffffff` reads identically through `getBigInt64` and `getBigUint64`, so it cannot distinguish signed from unsigned. The big-endian block on line 91 does pin signedness with `-9223372036854775808n`, and both methods share the one `this.little` flag rather than having separate paths, so there is no realistic bug hiding here - but a negative value (e.g. `ffffffffffffffff` → `-1n`) in the little-endian half would make the assertion carry its weight. Same for `readUInt64`, whose little-endian value is also below 2^63.
